drivers: pcie: Add support for IRQ allocation management

jhedberg opened the pull request:
There are x86 platforms where the IRQ configuration register for PCIe
is not pre-populated and the OS needs to assign a number dynamically
by writing to the register.

In order to allocate interrupts we have to know which ones have been
hard-coded in device tree. We accomplish this by collecting these
values through the IRQ_CONNECT() macro and placing them in a dedicated
linker section (in ROM).

The full set of allocated interrupts are managed through a bitmap, and
the pre-allocated values (from the linker section) are inserted into
this upon initial runtime access.

The new API for allocating IRQs for PCIe is called `pcie_alloc_irq()` and the
existing `pcie_wired_irq()` is renamed to `pcie_get_irq()` to better reflect
its purpose.

Commit:
drivers: pcie: Add support for IRQ allocation management

There are x86 platforms where the IRQ configuration register for PCIe
is not pre-populated and the OS needs to assign a number dynamically
by writing to the register.

In order to allocate interrupts we have to know which ones have been
hard-coded in device tree. We accomplish this by collecting these
values through the IRQ_CONNECT() macro and placing them in a dedicated
linker section (in ROM).

The full set of allocated interrupts are managed through a bitmap, and
the pre-allocated values (from the linker section) are inserted into
this upon initial runtime access.

This patch introduces a new pcie_alloc_irq() API that drivers can use
to allocate interrupt line numbers. The two in-tree drivers that were
using this API (I2C and UART) are converted to use the new API.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -68,11 +68,78 @@ uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
+		data &= ~0xffU;
+		data |= irq;
+		pcie_conf_write(bdf, PCIE_CONF_INTR, data);
+	} else {
+		atomic_set_bit(irq_reserved, irq);
+	}
```
I'm considering removing this else-branch. It's there to prevent conflicts if some devices have a pre-set IRQ but others don't, however I haven't seen such "mixed" platforms yet. The consequence of setting the bit here is that multiple calls to `pcie_wired_irq()` for the same device would each time allocate a new IRQ numbers, which would mess things up. This would happen e.g. if you repeatedly call `pcie show` in the PCIe shell.

jhedberg replied:
Another solution would be to split things into two APIs: one for allocating and one for getting the existing value. Or, add a parameter to `pcie_wired_irq()` to request whether we want to try allocation or do a read-only operation. Either of these will be more intrusive since they will then require changes to existing drivers.

tbursztyka replied:
splitting the logic makes sense. There isn't much drivers using pcie_wired_irq (only 2 actually)

tbursztyka replied:
About the split:
drivers would call pcie_allocate_irq(bdf). 
Under the hood it would get the irq, if proper would check if not allocated already, etc... etc...
So pcie_get_irq() could still be used by the PCIe shell and would not have this border effect of reallocating

dcpleung commented on drivers/pcie/host/pcie.c:
```diff
@@ -68,11 +68,78 @@ uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
+			fz--; /* find_lsb_set returns index + 1 */
+
+			if (!atomic_test_and_set_bit(&irq_reserved[i], fz)) {
+				return fz + (i * sizeof(atomic_val_t) * 8);
```
Might want to make sure the result is less than `CONFIG_MAX_IRQ_LINES`. Or else, this might return 5 when `CONFIG_MAX_IRQ_LINES==4`, since `find_lsb_set()` sees bit 5 as set in the above while condition (assuming atomic_val_t is at least an `uint8_t`).

jhedberg replied:
That's a good point, the way `irq_reserved` is defined it will fit bits rounded up to the next multiple of `(sizeof(atomic_val_t) * 8`, which might be more than `MAX_IRQ_LINES`. I'll fix it.

dcpleung commented on include/arch/x86/intel64/arch.h:
```diff
@@ -93,11 +93,20 @@ struct x86_ssf {
  */
 
 #define ARCH_IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p) \
+	X86_RESERVE_IRQ(irq_p, _CONCAT(_irq_alloc_fixed, __COUNTER__)); \
```
IIRC, `__COUNTER__` starts as 0 for each source file. Multiple drivers may result in multiple `_irq_alloc_fixed0` (and so on).

jhedberg replied:
That's a fair point, but it appears it's not an issue since we never use these variables by name but by array index. I managed to link an image where both HPET and GPIO had the same `_irq_alloc_fixed0` name without any compiler or linker issues:
```
irq_alloc       0x000000000010b130        0x2                                   
                0x000000000010b130                __irq_alloc_start = .         
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(._irq_alloc*)))                               
 ._irq_alloc.static._irq_alloc_fixed0                                           
                0x000000000010b130        0x1 zephyr/libzephyr.a(hpet.c.obj)    
 ._irq_alloc.static._irq_alloc_fixed0                                           
                0x000000000010b131        0x1 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_intel_apl.c.obj)
                0x000000000010b132                __irq_alloc_end = .   
```

So I'd conclude that it's only important that these are unique within each compilation unit (which `__COUNTER__` does guarantee).

jhedberg commented on include/linker/common-rom.ld:
```diff
@@ -138,6 +138,15 @@
 	Z_ITERABLE_SECTION_ROM(dns_sd_rec, 4)
 #endif
 
+#if defined(CONFIG_PCIE)
```
I made it for `CONFIG_PCIE` since that way we don't need #ifdefs in `drivers/pcie/host/pcie.c` (since the entire file itself is conditioned on `CONFIG_PCIE=y`). In the worst case non-X86 platforms supporting PCIe might just become slightly less broken since they will attempt allocation of IRQs for devices that do not have one pre-assigned. The only thing they won't have (until it's implemented) is the support through the `IRQ_CONNECT()` macro, which means that there could be collisions, however to my understanding that's primarily an efficiency issue (executing ISRs when not necessary).

andyross replied:
Strictly yes: there are devices out there with PCI buses that aren't x86.  None that Zephyr supports, but there's no point in putting in needless architecture dependencies.

Commit:
drivers: pcie: Rename pcie_wired_irq to pcie_get_irq

This reflects better the actual purpose of the API.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@dcpleung @tbursztyka I've made changes based on the feedback, please take another look.

Commit:
drivers: serial: ns16550: Update to use new pcie_alloc_irq API

Use the new pcie_alloc_irq() API so that we get a valid IRQ on
platforms where the IRQ registers do not contain valid values up
front.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
drivers: i2c: dw: Update to use new pcie_alloc_irq API

Take advantage of the new pcie_alloc_irq() API so that we get a valid
IRQ on platforms where the IRQ register is not pre-populated with a
valid value up front.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

tyrantlizard77 reviewed (commented):
I don't think I fully understand this patch.
Shouldn't we be using `irq_connect_dynamic()`?
On x86_64, all interrupts are dynamic anyway...

jhedberg commented:
> Shouldn't we be using irq_connect_dynamic()?

@andrewboie `irq_connect_dynamic()` assumes that you have an IRQ number up front, however the case with some newer x86 platforms is that we don't have that (neither in device tree nor in the PCIe configuration registers at runtime) so we need a way of allocating numbers and writing them to the PCIe configuration. It's after all that that `irq_connect_dynamic()` comes into play, i.e. it's really about a different phase of initialization/configuration.

andrewboie commented:
Would value @andyross review on this one, but LGTM.

andyross commented on drivers/pcie/host/pcie.c:
```diff
@@ -68,6 +68,83 @@ uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
+ */
+#define IRQ_LIST_INITIALIZED 0
+
+static ATOMIC_DEFINE(irq_reserved, CONFIG_MAX_IRQ_LINES);
```
Why atomic_t and not just integers?  Seems like this is only mutated from initialization code that could easily be locked.  Though I guess it's not much savings in code size as the atomic bit API exists.

jhedberg replied:
It was mainly not to reimplement bit manipulation APIs. ATOMIC_DEFINE seemed also convenient since it takes a number of needed bits and gives you a sufficiently large array. But you’re right that it’s perhaps a bit overkill since access is limited to initialization code.

andyross commented on include/arch/x86/intel64/arch.h:
```diff
@@ -93,11 +93,20 @@ struct x86_ssf {
+#ifdef CONFIG_PCIE
+#define X86_RESERVE_IRQ(irq_p, name) \
+	static Z_DECL_ALIGN(uint8_t) name \
+	__in_section(_irq_alloc, static, name) __used = irq_p
```
Seems like this should be using the Z_ITERABLE API, no?   It's just an array of static structs to be enumerated, right?

jhedberg replied:
I’d have wanted to use that, but it’s not structs here, rather plain uint8_t entries. I guess I could have wrapped them in structs but that felt a bit overkill.

andyross reviewed (approved):
Some style/usage questions/nits.  Nothing that should block merge.
